src/vcpkg/dependencies.cpp

```diff
@@ -1300,36 +1300,43 @@ namespace vcpkg::Dependencies
                 std::vector<std::string> features;
             };
 
-            // This object contains the current version within a given column.
-            // Each "string" scheme text is treated as a separate column
-            // "relaxed" versions all share the same column
+            struct PackageNode;
+            // This object contains the current version within a given version sheme (except for the "string" scheme,
+            // there we save an object for every version)
             struct VersionSchemeInfo
             {
                 Versions::Scheme scheme;
                 const SourceControlFileAndLocation* scfl = nullptr;
                 Versions::Version version;
                 // This tracks a list of constraint sources for debugging purposes
                 std::vector<std::string> origins;
+                // mapping from feature name -> dependencies of this feature
                 std::map<std::string, std::vector<FeatureSpec>> deps;
+                std::set<PackageNode*> dependencies; // track to which PackageNodes this vsi has dependencies
 
                 bool is_less_than(const Versions::Version& new_ver) const;
             };
 
             struct PackageNode
             {
+                // Mapping from version to the newest version in the cooresponding version sheme
                 std::map<Versions::Version, VersionSchemeInfo*, VersionTMapLess> vermap;
+                // We don't know how to compare "string" versions, so we save all versions seperatly
                 std::map<std::string, VersionSchemeInfo> exacts;
+                // Within each version sheme("version", semver, date) we only track the newest version
                 Optional<std::unique_ptr<VersionSchemeInfo>> relaxed;
                 Optional<std::unique_ptr<VersionSchemeInfo>> semver;
                 Optional<std::unique_ptr<VersionSchemeInfo>> date;
-                std::set<std::string> features;
-                bool default_features = true;
+                std::set<std::string> requested_features;
+                // state to track if default features must be enables at next add_port_at_version call
+                bool update_default_features = false;
                 bool user_requested = false;
 
                 VersionSchemeInfo* get_node(const Versions::Version& ver);
+                // Created/Updates(is newer than existing) the version entry for the given sheme
                 VersionSchemeInfo& emplace_node(Versions::Scheme scheme, const Versions::Version& ver);
 
-                PackageNode() = default;
+                PackageNode(bool only_host_dependency) : only_host_dependency(only_host_dependency){};
```

Comment:
```suggestion
                explicit PackageNode(bool only_host_dependency) : only_host_dependency(only_host_dependency){};
```
This should also use an `enum class`.

Reply:
In general I agree, but this would make the implementation very verbose. For Example `only_host_dependency && host_dependency` would be `((only_host_dependency == OnlyHostDependency::Yes) && (host_dependency == OnlyHostDependency::Yes)) ? OnlyHostDependency::Yes : OnlyHostDependency::No` which seems very verbose.

Reply:
@strega-nil-ms Any comment?